### ISSUE TITLE
Created WaterBody class to consolidate code from Pool and Stream classes

### ISSUE THREAD

**Issue by paulszefer** [paulszefer/Ecosystem-Simulation#2]
Hi Chris, I added a WaterBody class to collect similar code from the Pool and Stream classes. It is on the branch 'new/WaterBody'. Am I on the right track with your suggestion to shorten our classes?

**Comment by chris-thompson** [comment]
Very nice.  Yes, I think inheritance is definitely one of the strategies you should be using.  I like that you've extracted the common attributes and pushed them as 'far' up the tree as they can go.

I am curious whether you think that there is anywhere we could make use of interfaces... :-)

**Comment by paulszefer** [comment]
So far I have come across two possible uses for interfaces in our simulation:

I read about programming to interfaces instead of to specific types. For this I would change some of the ArrayList declarations to be Lists instead. This would not change any function, but it would allow us to refactor any of the fields using ArrayLists to use another type of List instead.

I could also add interfaces to define the essential functionality of our project. This would be something like an IHabitat interface to represent the habitat of the creatures (implementing ICreature) that we want to represent. These interfaces would declare the necessary functionality for our simulation like adding a creature to the habitat, incrementing the ages of the creatures in the habitat, etc.

This would allow us to create simulations of other creatures and habitats as well; however, I don't see how that would shorten any of the current classes.

I can't find any resources talking about how Interfaces can consolidate code (other than a suggestion to create an Interface to store constants, which was subsequently recommended against).

**Comment by chris-thompson** [comment]
Yes!  Programming to an interface is something every Java developer does.  I'm glad you found this.  I would have told you eventually 👍 

How about the Guppies or other creatures themselves?  Is there any way we can take subsets of these classes and move them into subclasses?

**Comment by paulszefer** [comment]
I could definitely make more general classes for Guppy to extend such as a fish class, then an aquatic creature class, then a creature class, etc. I'll have to see which of those would work best, if not all of them.

Did you mean to say superclasses instead of subclasses?

**Comment by chris-thompson** [comment]
Not quite.  I'm thinking of something a bit different.  For example, instead of storing 2 strings for name and surname in a Student class, we will often create a Name class that stores a first and last name, and then use a Name instance inside the Student.  I don't have anything in particular in mind, but I am curious whether you observe any such "encapsulatable" subset of attributes in any of our current classes.

**Comment by paulszefer** [comment]
Ok, I understand what you are suggesting.
There are a few things that I notice in the Guppy class.
- the guppy has static (non-changing) state (species, gender, identification number)
- the guppy has mutable state (its health, age, whether it is alive)

The first could be consolidated into descriptive/identifying characteristics, while the second could be grouped in a representation of the guppy's health.

**Comment by chris-thompson** [comment]
Nice work!  I would enjoy seeing you investigate this!

**Comment by paulszefer** [comment]
Thank-you! I will see what I can do with those options.

**Comment by paulszefer** [comment]
Hi Chris,
I have simplified Guppy to contain an identification profile (containing a classification profile (containing a genus and species), an identification number, and a generation number), a health profile (containing whether the guppy is alive, its age, and its health coefficient).

I have now moved onto the concept of programming to an interface. I changed the declarations of ArrayLists to be Lists instead. I am now refactoring the program to handle 'creatures' (using a Creature interface) instead of guppies. For the most part, this is fine; however, I have run into a couple of issues:
1. Static methods cannot be abstract; I can't force implementing classes to contain a 'getNumberBorn()' method by adding a static version in the interface. This is not needed yet, but it might be in the future.
2. How can I dynamically populate the pool based on a loaded data file? The only way that I can think of is to hard-code a check such as:
`String creatureType = data.nextLine();`
`if (creatureType.equals("Guppy")) {`
`    // populate the habitat with guppies according to the following information`
`} else if (creatureType.equals("Turtle")) {`
`    // populate with turtles`
`}`

**Comment by chris-thompson** [comment]
Great comments.  Yes, statics cannot be abstract and as we study abstracts and the class loader in 2526 you will understand why.  We will also study reflection which I think will help with your second question.  I'm glad you're thinking about these things!  Check out reflection if you have time...
